app/slave/cluster_slave.py

```diff
@@ -173,18 +175,20 @@ def _do_build_teardown_and_reset(self, timeout=None):
         for executor in self.executors_by_id.values():
             executor.kill()
 
-        if not self._project_type:
-            return  # There is no build to tear down! (Slave is idle.)
+        # This lock keeps multiple teardown attempts from racing each other and throwing exceptions.
+        with self._build_teardown_lock:
```

Comment:
I think the original intent of the build_teardown_coin was to signal that this code has no legitimate scenario where it can be executed multiple times for the same build. But now we have just introduced a legitimate scenario for it.

I'd say go with 1). as well. We definitely don't need both though.